### PR TITLE
QUIC Conformance Updates: Key Update

### PR DESCRIPTION
**Pull request by hlandau**
QUIC conformance updates in the area of key updates. This clears the conformance board for RFC 9001 s. 6.

Built on top of the SSL_tick rename PR.

```
15e290d546 QUIC: Test key update works correctly
57a97dc129 QUIC CHANNEL: Optimise key update using ACKs
b505ba2a5c QUIC: Add internal APIs for white-box testing of key update
26c5b45834 QUIC CHANNEL: Enforce the RX packet forgery limit
e717de7400 QUIC RXDP: Strictly enforce ACK PNs with regard to TX key epochs
8e62246a35 QUIC CHANNEL: Ensure new packets aren't enforced with old keys
1678db040c QUIC CHANNEL: Handle key updates correctly
43fc073ae3 QUIC TXP: Make TXP use time callback correctly
f27a84c5ed QUIC APL: Correct implementation of time callback override
32753e6d06 QUIC RX: Support reporting the key epoch a packet was received with
fc29444df2 QUIC TXP: Allow callbacks on ACK transmission
57465179d7 QUIC TXP: Allow next PN to be used to be queried
f0bb2ee9a9 QUIC RX: Refactor key update callback to provide PN
5eb2f0c05e QUIC TXP: Remove TX key update handling from TXP
a856e12317 QUIC TXP: Refactor status output to use an extensible structure
22532dd043 QUIC ACKM: Allow largest acked PN to be queried
8498334fa8 QUIC WIRE: Utility function to determine if PN is in an ACK frame
02ca5a2382 QUIC Glossary: TXKU, RXKU, KU
```

### CODE REVIEW

**Comment by hlandau** [comment]
Rebased

**Comment by hlandau** [comment]
Rebased

File: ssl/quic/quic_channel.c [hunk 1]
```diff
@@ -184,6 +187,7 @@ static int ch_init(QUIC_CHANNEL *ch)
                                   ch->cc_method, ch->cc_data)) == NULL)
         goto err;
 
+
```
Comment (mattcaswell) [inline ssl/quic/quic_channel.c:190]: Spurious blank line inserted here

File: ssl/quic/quic_channel.c [hunk 2]
```diff
@@ -512,6 +524,235 @@ static int rx_early_validate(QUIC_PN pn, int pn_space, void *arg)
...
+     * which it more or less should be; it is necessarily separate from ordinary
+     * processing of ACK-eliciting frames as key update is not indicated via a
+     * frame.)
+     */
+    ossl_quic_tx_packetiser_schedule_ack(ch->txp, QUIC_PN_SPACE_APP);
+}
+
+/* Called per tick to handle RXKU timer events. */
+QUIC_NEEDS_LOCK
+static void ch_rxku_tick(QUIC_CHANNEL *ch)
```
Comment (mattcaswell) [inline ssl/quic/quic_channel.c:725]: Is tick the right name here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_record_rx.c [hunk 3]
```diff
@@ -678,13 +740,14 @@ static ossl_inline void ignore_res(int x)
     /* No-op. */
 }
 
-static void qrx_key_update_initiated(OSSL_QRX *qrx)
+static void qrx_key_update_initiated(OSSL_QRX *qrx, QUIC_PN pn)
 {
     if (!ossl_qrl_enc_level_set_key_update(&qrx->el_set, QUIC_ENC_LEVEL_1RTT))
+        /* Returns 0 if already in RXKU, so we don't call callback again. */
         return;
```
Comment (t8m) [inline ssl/quic/quic_record_rx.c:747]: I assume `returns 0` in the comment refers to the ossl_qrl_enc_level_set_key_update() function return. Would it make sense to change the comment to : `We are already in RXKU, so we don't call callback again.`

**Review by t8m — CHANGES_REQUESTED** [review]
Just very minor nits. Good work!

File: ssl/quic/quic_channel_local.h [hunk 4]
```diff
@@ -193,6 +193,14 @@ struct quic_channel_st {
      */
     uint64_t                        incoming_stream_auto_reject_aec;
 
+    /*
+     * Override packet count threshold at which we do a spontaneous TXKU.
+     * Usually UINT64_MAX in which case a suitable value is chosen based on AEAD
+     * limit advice from the QRL utility functions. This is intended for testing
+     * use only. Usually set to UINT64_MAX.
+     */
+    uint64_t                        txku_threshold_override;
```
[the remark this answers is not in the document]
Comment (hlandau) [reply ssl/quic/quic_channel_local.h:202]: Good point, I'll wire this up.

**Comment by hlandau** [comment]
Updated.

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Review by paulidale — APPROVED** [review]

**Pull request closed without merging**